File: frontend/src/api/jam.ts
import { Id, send } from "../client";

interface Team {
  lead: boolean;
  lost: boolean;
  starPas: number | null;
  trips: { timestamp: string; points: number }[];
  jammer: null;
  blockser: [null, null, null, null];
  noPivot: boolean;
}

interface Jam {
  startTimestamp: string;
  stopTimestamp: string;
  stopReason: string;
  home: Team;
  away: Team;
}

const jams: Map<Id, Jam> = new Map();


export async function getJam(boutId: string, periodId: number,
  jamId: number): Promise<Jam> {
  const jamKey: Id = { boutId, periodId, jamId };
  if (jams.has(jamKey)) {
    return <Jam>jams.get(jamKey);
  }
  const jam = <Jam>await send("getJam", jamKey);
  jams.set(jamKey, jam);
  return jam;
}

File: backend/src/model/bouts/jam.py
from dataclasses import dataclass, field
from datetime import datetime
from typing import TYPE_CHECKING, Final, Literal
from uuid import UUID

from .attribute import TeamAttribute, QueryKey, Queryable

type JamId = tuple[int, int]
type JamStopReasons = Literal['called', 'time', 'injury', 'other']


@dataclass(slots=True)
class TripState:
    points: int
    timestamp: datetime


@dataclass(slots=True)
class ScoreState:
    lead: bool = False
    lost: bool = False
    star_pass: int | None = None
    trips: Final[list[TripState]] = field(default_factory=list)


@dataclass(slots=True)
class JamState(TeamAttribute[ScoreState],):
    start_timestamp: datetime | None = None
    stop_timestamp: datetime | None = None
    stop_reason: JamStopReasons | None = None
    home: Final[ScoreState] = field(default_factory=ScoreState)
    away: Final[ScoreState] = field(default_factory=ScoreState)

    def __getitem__(self, key: Literal['home', 'away']) -> ScoreState:
        if key == 'home':
            return self.home
        if key == 'away':
            return self.away
        raise KeyError(f"Invalid team: {key}")

    def lead_is_declared(self) -> bool:
        return self.home.lead or self.away.lead

    def get_query_key(self, bout_id: UUID, jam_id: JamId) -> QueryKey:
        return (bout_id, 'jam', jam_id)
